Add client.wait method

knkski opened the pull request:
Adds `client.wait` as a rough analogue of `kubectl wait`. Builds on top of `client.watch` to wait for specified resource to enter expected state.

Opening as draft PR to ensure that this is something that is generally desired in lightkube. If so, will also implement logic for `AsyncClient` and convert to regular PR.

Punts on implementing an equivalent of `kubectl wait`'s `--timeout` flag, as that would require some deeper changes in how exactly the watch requests are handled. Some extra logic around [`stream=True`](https://github.com/gtsystem/lightkube/blob/436126f/lightkube/core/generic_client.py#L211) and setting the [read timeout](https://github.com/gtsystem/lightkube/blob/436126f/lightkube/core/generic_client.py#L80) should work if that's desired, but this PR also includes an example of how to achieve the same thing with `gevent.Timeout`.

gtsystem commented:
I think adding this functionality is a nice idea, as it's not straight forward to implement. 

Few things I was thinking about regarding the interface:
* It would be nicer to align the naming more with kubectl, so instead of `success_conditions` maybe `for_conditions`.
* In the future we could also add support for "wait for delete" using `for_delete=True`.
* Failure cases are not supported by kubectl but I think it's a good addition, we could name it `raise_for_conditions` (this clarify that the effect is an exception and is similar to `requests` method `raise_for_status` ).
* Defaults: I'm not sure providing this defaults are a good idea:
  1. Success conditions can vary across different resources and new ones can come in the future, once we define this default `DEFAULT_SUCCESS_CONDITIONS` we cannot modify it as it would be a breaking change. I would prefer this parameter to be required instead (as in kubectl wait)
  2. Checking for failures may not be always wanted and to disable this default the user will be force to add `raise_for_status=[]` which is not ideal. Empty should be the default instead.
* kubectl supports passing the condition status to verify (true or false). Not sure how useful this is but we could support this in the future using a dict like value instead `for_conditions={"Available": False}`. 
* timeout: This is something we may want to support but we could go without it for the first version.

Any opinion/idea?

gtsystem commented on lightkube/core/client.py:
```diff
@@ -176,6 +179,76 @@ def watch(self, res, *, namespace=None, labels=None, fields=None, server_timeout
+        failure_conditions = list(failure_conditions)
+
+        for op, obj in self.watch(res, namespace=namespace):
+            if obj.metadata.name != name:
```
We can use a field selector in the watch call instead matching `metadata.name`

gtsystem commented on lightkube/core/client.py:
```diff
@@ -176,6 +179,76 @@ def watch(self, res, *, namespace=None, labels=None, fields=None, server_timeout
+
+            conditions = list(sorted(status['conditions'], key=lambda c: c['lastTransitionTime']))
+
+            # There may be multiple conditions with the same timestamp, so grab all with the most
```
I think we should look at all conditions:
1. Maybe the condition did happen already before we exec the watch
2. Conditions can be modified by different components running on different nodes with different internal clock (for example kubelet and an operator). As a result picking the last `lastTransitionTime` may not always be accurate.

knkski commented:
Thanks for the review, and sorry for the slow response. I've updated this branch with all requested changes other than the timeouts and `for_conditions` supporting a dict-like value. I've also added `AsyncClient.wait` to mirror `Client.wait`.

gtsystem commented on lightkube/core/client.py:
```diff
@@ -176,6 +176,69 @@ def watch(self, res, *, namespace=None, labels=None, fields=None, server_timeout
+                continue
+
+            if op == "DELETED":
+                raise Exception(f"{kind}/{name} was deleted while waiting for it")
```
let's define and use a custom exception `ObjectDeleted`(?) in `core.exceptions`.

gtsystem commented on lightkube/core/client.py:
```diff
@@ -176,6 +176,69 @@ def watch(self, res, *, namespace=None, labels=None, fields=None, server_timeout
+
+            if failures:
+                messages = '; '.join([f.message for f in failures])
+                raise Exception(f"Error while waiting for {kind}/{name}: {messages}")
```
Custom exception here as well: `ConditionError` or `ConditionException`?

gtsystem commented on lightkube/core/client.py:
```diff
@@ -176,6 +176,69 @@ def watch(self, res, *, namespace=None, labels=None, fields=None, server_timeout
+        name: str,
+        *,
+        namespace=None,
+        for_conditions: Iterable[str] = (),
```
I guess you put the parameter here and optional for consistency, but in this way your editor (or pylint) won't show an error if this parameter is not provided. I think it's fine to have it required but we can force it to be a keyword argument so that position won't matter as far as it is defined.

gtsystem replied:
Reopening this one, I was suggesting to force it to be a keyword argument (so appear after the '*' ) but right now it seems to be only positional. The reason is that if in the future we want to add `for_delete`, we will need to bring `for_conditions` back as optional (need to be mutually exclusive with `for_delete`) and we want to be sure we can do that in a backward compatibility way.

gtsystem replied:
I think after this we are good to merge 👍

knkski replied:
Ah sure, sounds good. Updated

knkski replied:
Whoops, missed a spot, as I was accidentally filtering the tests to be ran from `e2e-tests/` when running locally. Let me update that.

knkski replied:
Alright, updated. Sorry about that

gtsystem commented on lightkube/core/client.py:
```diff
@@ -176,6 +176,69 @@ def watch(self, res, *, namespace=None, labels=None, fields=None, server_timeout
+            except AttributeError:
+                status = obj.status
+
+            if any(c['type'] in for_conditions for c in status['conditions']):
```
We should filter for `c['status'] == 'True'`. Same in line 236.

gtsystem commented on lightkube/core/client.py:
```diff
@@ -176,6 +176,69 @@ def watch(self, res, *, namespace=None, labels=None, fields=None, server_timeout
+        for_conditions: Iterable[str] = (),
+        raise_for_conditions: Iterable[str] = (),
+    ):
+        """Waits for specified conditions."""
```
Missing documentation on the parameters.

gtsystem commented on e2e-tests/test_client.py:
```diff
@@ -1,11 +1,15 @@
+from gevent import monkey, Timeout
```
Can we remove the depenency on gevent?

knkski replied:
I was including it to show how someone can set a custom time limit on `client.wait`. Removed it in favor of eventually adding a `timeout` parameter to `client.wait`.

gtsystem reviewed (commented):
Code coverage for the two new functions is missing. Tests should go in `tests/test_client.py` and `tests/test_async_client.py`. Be sure we are covering every condition.

knkski commented on tests/test_async_client.py:
```diff
@@ -9,9 +9,12 @@
 from lightkube.models.meta_v1 import ObjectMeta
 from lightkube import types
 
+from .test_client import make_wait_custom, make_wait_deleted, make_wait_failed, make_wait_success, make_watch_list
```
Added `tests/__init__.py` so that we can just import as `from .test_client` instead of `from test.test_client` and needing to add a custom path.

gtsystem replied:
ok

gtsystem commented on e2e-tests/test_client.py:
```diff
@@ -246,4 +247,62 @@ def test_list_all_ns(obj_name):
+@pytest.mark.parametrize(
+    "resource,for_condition", [(Pod, "Ready"), (Deployment, "Available")]
+)
+def test_wait_namespaced(resource, for_condition):
```
This test seems to fail randomly (see CI) or take a lot of time (condition never met?). I suggest to wait for an object under control of the test. For example we can create a pod (like in other tests) and wait for it to be ready using client.wait().

knkski replied:
@gtsystem Good call, updated. That also caught a bug where `status.conditions` can be empty if you create an object then wait for it right away.

gtsystem commented:
Merged. @knkski, thanks a lot for the contribution :)
